rfcs/0000-ckb-light-client/0000-ckb-light-client.md

```diff
@@ -0,0 +1,507 @@
+---
+Number: "0000"
+Category: Standards Track
+Status: Proposal
+Author: Boyu Yang <yangby@cryptape.com>
+Created: 2022-08-18
+---
+
+# CKB Light Client Protocol
+
+## Abstract
+
+This RFC describes a light client protocol on CKB which allows clients to
+verify that a blockchain is valid with limited resources.
+
+## Motivation
+
+Downloading and verifying all blocks is taking hours and requiring gigabytes
+of bandwidth and storage.  Hence, clients with limited resources cannot
+verify transactions independently without trusting full nodes.
+Even some light clients only downloads  all block headers, the storage and
+bandwidth requirements of those clients still increase linearly with the
+chain length.
+So CKB offers a solution base on [FlyClient] which is an efficient light
+client for chains of variable difficulty. It requires downloading only a
+logarithmic number of block headers while storing only a single block header
+between executions.
+
+## Background
+
+### Merkle Mountain Range (MMR)
+
+A [Merkle Mountain Range (MMR)][MMR] is a binary hash tree that allows for
+efficient appends of new leaves without changing the value of existing
+nodes.
+
+In MMR, we use the insertion order to reference leaves and nodes.
+We insert a new leaf to MMR as following steps:
+- Insert leaf or node to next position.
+- If the current position has a left sibling, we merge the left and right
+  nodes to produce a new parent node, then go back to step 1 to insert the
+  node.
+
+For example, we insert a leaf to the example MMR which has 11 leaves as
+following steps:
+- Insert leaf to next position: `19`.
+- Then check the left sibling `18` and calculate parent node:
+  `merge(mmr[18], mmr[19])`.
+- Insert parent node to position `20`.
+- Since the node `20` also has a left sibling `17`, calculate parent node:
+  `merge(mmr[17], mmr[20])`.
+- Insert new node to next position `21`.
+- Since the node `21` have no left sibling, complete the insertion.
+
+```
+# An MMR with 11 leaves:
+          14
+       /       \
+     6          13
+   /   \       /   \
+  2     5     9     12     17
+ / \   /  \  / \   /  \   /  \
+0   1 3   4 7   8 10  11 15  16 18
+
+# After insertion of a new leaf:
+
+          14
+       /       \
+     6          13            21
+   /   \       /   \         /   \
+  2     5     9     12     17     20
+ / \   /  \  / \   /  \   /  \   /  \
+0   1 3   4 7   8 10  11 15  16 18  19
+```
+
+#### Merkle Root
+
+An MMR is constructed by one or more sub merkle trees (or mountains). Each
```

Comment:
It's better to give a formal definition of how to *bag* the peaks. For the example, it's hard to deduce how to compute the root containing 4 peaks.